### PR TITLE
added AppendBuf and got all doctests passing

### PR DESCRIPTION
**Pull request by rrichardson**

### CODE REVIEW

File: src/appendbuf.rs [hunk 1]
```diff
@@ -0,0 +1,460 @@
...
+    ///   let middle = buf.atomic_slice_pos_from_end(16, 8).ok().expect("pos_from_end");
+    ///   let end = buf.atomic_slice_from_end(8).ok().expect("from_end");
+    ///   let meh = buf.atomic_slice_pos_from_begin(4, 8).ok().expect("pos_from_begin");
+    ///
+    ///   let b = unsafe { begin.as_window_slice() };
+    ///   let m = unsafe { middle.as_window_slice() };
+    ///   let e = unsafe { end.as_window_slice() };
+    ///   let z = unsafe { meh.as_window_slice() };
+    ///
+    ///   assert_eq!(b, ['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H'].iter().map(|&: x| *x as u8).collect::<Vec<u8>>());
```
Comment (cgaebel) [inline src/appendbuf.rs]: Couldn't these be replaced with `b"ABCDEFGH".iter()`

File: src/appendbuf.rs [hunk 2]
```diff
@@ -0,0 +1,460 @@
...
+    /// buffer can continue to be filled and poked. There are no operations for this buffer
+    /// to reset the window to a lower position in the buffer.
+    /// pos is the offset backwards from the beginning of the window
+    /// len is the length of the slice
+    ///
+    /// This function does not mutate the buffer, window, or limits.
+    /// It just increases the refcount.
+    ///
+    /// ```rust
+    /// #![allow(unstable)]
```
Comment (cgaebel) [inline src/appendbuf.rs]: Ideally none of the doctests would have `#![allow(unstable)]`. I don't see why Iobuf can't work with rust stable when it comes out.

File: src/appendbuf.rs [hunk 3]
```diff
@@ -0,0 +1,460 @@
...
+    ///   let e = unsafe { end.as_window_slice() };
+    ///   let z = unsafe { meh.as_window_slice() };
+    ///
+    ///   assert_eq!(b, ['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H'].iter().map(|&: x| *x as u8).collect::<Vec<u8>>());
+    ///   assert_eq!(m, ['I', 'J', 'K', 'L', 'M', 'N', 'O', 'P'].iter().map(|&: x| *x as u8).collect::<Vec<u8>>());
+    ///   assert_eq!(e, ['Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X'].iter().map(|&: x| *x as u8).collect::<Vec<u8>>());
+    ///   assert_eq!(z, ['E', 'F', 'G', 'H', 'I', 'J', 'K', 'L'].iter().map(|&: x| *x as u8).collect::<Vec<u8>>());
+    /// ```
+    ///
+    #[inline(always)]
```
Comment (cgaebel) [inline src/appendbuf.rs]: Maybe just `#[inline]`. I generally reserve `#[inline(always)]` for things that I _know_ will compile down to less code than a `call`

File: src/appendbuf.rs [hunk 4]
```diff
@@ -0,0 +1,460 @@
+use std::fmt::{self, Show, Formatter};
+use std::mem;
+use std::num::Int;
+use std::sync::Arc;
+
+use raw::{Allocator, RawIobuf};
+use iobuf::{Iobuf};
+use impls::{AROIobuf};
+
+/// ## AppendBuf
```
Comment (cgaebel) [inline src/appendbuf.rs]: This doesn't match the style of the other impls. The first line should be "Append-only input buffer"

File: src/appendbuf.rs [hunk 5]
```diff
@@ -0,0 +1,460 @@
...
+/// This buffer is intended to act as a intermediate buffer to be filled by
+/// incoming streaming data, such as sockets, files, or perhaps DB results
+///
+/// It has the unique feature of being able to break off reference counted
+/// slices of data in its buffer that it has already written.
+///
+/// This invariant is enforced by allowing slices to be taken only from the
+/// low side of the buffer, before the start of the window
+///
+/// Its primary interface is fill, which is the mechanism for appending data,
```
Comment (cgaebel) [inline src/appendbuf.rs]: code (like "fill") should be in backticks

File: src/appendbuf.rs [hunk 6]
```diff
@@ -0,0 +1,460 @@
...
+/// and atomic_slice, which will take a position
+/// and a length and return a Result<AROIobuf, ()>
+///
+#[unsafe_no_drop_flag]
+pub struct AppendBuf<'a> {
+    raw: RawIobuf<'a>,
+}
+
```
Comment (cgaebel) [inline src/appendbuf.rs]: nit: extra space

File: src/appendbuf.rs [hunk 7]
```diff
@@ -0,0 +1,460 @@
...
+    ///
+    /// This function does not mutate the buffer, window, or limits.
+    /// It just increases the refcount.
+    ///
+    /// ```rust
+    /// #![allow(unstable)]
+    /// use iobuf::{AppendBuf, Iobuf};
+    ///
+    ///   let mut buf = AppendBuf::new(24);
+    ///   for i in b'A' .. b'X' + 1 {
```
Comment (cgaebel) [inline src/appendbuf.rs]: isn't `...` inclusive range? or does that not exist?

File: src/appendbuf.rs [hunk 8]
```diff
@@ -0,0 +1,460 @@
...
+    ///   assert_eq!(b, ['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H'].iter().map(|&: x| *x as u8).collect::<Vec<u8>>());
+    ///   assert_eq!(m, ['I', 'J', 'K', 'L', 'M', 'N', 'O', 'P'].iter().map(|&: x| *x as u8).collect::<Vec<u8>>());
+    ///   assert_eq!(e, ['Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X'].iter().map(|&: x| *x as u8).collect::<Vec<u8>>());
+    ///   assert_eq!(z, ['E', 'F', 'G', 'H', 'I', 'J', 'K', 'L'].iter().map(|&: x| *x as u8).collect::<Vec<u8>>());
+    /// ```
+    ///
+    #[inline(always)]
+    pub fn atomic_slice_pos_from_end(&self, pos: u32, len: u32) -> Result<AROIobuf, ()> { unsafe {
+        let mut ret = self.raw.clone_atomic();
+        let lim = (self.raw.lo() - pos, (self.raw.lo() - pos) + len);
```
Comment (cgaebel) [inline src/appendbuf.rs]: this can underflow if pos > self.raw.lo().

File: src/impls.rs [hunk 9]
```diff
@@ -108,14 +105,14 @@ impl<'a> Drop for RWIobuf<'a> {
 /// 0xFF, and consuming/validating these numbers in parallel in 4 other threads:
 ///
 /// ```rust
+/// #![allow(unstable)]
```
Comment (cgaebel) [inline src/impls.rs]: Does this example still need #[allow(unstable)], even with the range_inclusive change?

File: src/impls.rs [hunk 10]
```diff
@@ -198,44 +195,6 @@ impl Drop for AROIobuf {
   fn drop(&mut self) { unsafe { self.raw.drop_atomic() } }
 }
 
-impl Arbitrary for AROIobuf {
```
Comment (cgaebel) [inline src/impls.rs]: can we cfg(none) this instead?

File: src/impls.rs [hunk 11]
```diff
@@ -541,22 +504,23 @@ impl<'a> RWIobuf<'a> {
   /// multiple slices of the same data. Therefore, this function is unsafe.
   ///
   /// It may only be used safely if you ensure that the data in the iobuf never
-  /// interacts with the slice, as they point to the same data. `peek`ing or
+  /// interacts with the slice, as they pointe to the same data. `peek`ing or
```
Comment (cgaebel) [inline src/impls.rs]: thanks!

File: src/iobuf.rs [hunk 12]
```diff
@@ -423,7 +423,7 @@ pub trait Iobuf: Clone + Show {
   ///
   ///     let mut sum = 0u16;
   ///
-  ///     for _ in range(0, num_shorts) {
+  ///     for _ in (0 .. num_shorts) {
```
Comment (cgaebel) [inline src/iobuf.rs]: are the parentheses needed? not sure how this new syntax works

File: src/iobuf.rs [hunk 13]
```diff
@@ -457,7 +457,7 @@ pub trait Iobuf: Clone + Show {
 
   /// Returns `true` if the `other` Iobuf's window is the region directly after
   /// our window. This does not inspect the buffer -- it only compares raw
-  /// pointers.
+  /// pointeers.
```
Comment (cgaebel) [inline src/iobuf.rs]: well that doesn't look rigth

File: src/iobuf.rs [hunk 14]
```diff
@@ -755,9 +754,9 @@ pub trait Iobuf: Clone + Show {
   /// let mut tgt3 = [ 0x00, 0x00, 0x00 ];
   ///
   /// assert_eq!(b.peek(0, &mut tgt4), Ok(()));
-  /// assert_eq!(tgt4.iter().map(|&x| x).sum(), 10);
+  /// assert_eq!(tgt4.iter().map(|&x| x).fold(0,|a, b| a + b), 10);
```
Comment (cgaebel) [inline src/iobuf.rs]: Why?

File: src/raw.rs [hunk 15]
```diff
@@ -157,7 +157,7 @@ pub struct RawIobuf<'a> {
   // function.
   buf:    *mut u8,
   // If the highest bit of this is set, `buf` is owned and the data before the
-  // pointer is valid. If it is not set, then the buffer wasn't allocated by us:
+  // pointeer is valid. If it is not set, then the buffer wasn't allocated by us:
```
Comment (cgaebel) [inline src/raw.rs]: ?

File: src/raw.rs [hunk 16]
```diff
@@ -691,6 +691,24 @@ impl<'a> RawIobuf<'a> {
     Ok(())
   }
 
+  /// Both the limits and the window are [lo, hi).
+  #[inline]
+  pub fn expand_limits_and_window(&mut self, limits: (u32, u32), window: (u32, u32)) -> Result<(), ()> {
```
Comment (cgaebel) [inline src/raw.rs]: The difference between this and `set_limits_and_window` should be documented somewhere.

**Comment by cgaebel** [comment]
LGTM after nitpicks!

File: src/appendbuf.rs (hunk 1 above)
Comment (rrichardson) [inline src/appendbuf.rs]: Yea. That's probably a better idea.

File: src/appendbuf.rs (hunk 2 above)
Comment (rrichardson) [inline src/appendbuf.rs]: When it comes out. Right now its causing the tests to fail to compile.

File: src/appendbuf.rs (hunk 7 above)
Comment (rrichardson) [inline src/appendbuf.rs]: hmm.  According to https://github.com/rust-lang/rfcs/blob/master/text/0198-slice-notation.md that is, in fact, the case. Not sure why I didn't notice that before.  That seems like a horrible idea, but I can't think of a better one.  I'll adjust this.

File: src/appendbuf.rs (hunk 8 above)
Comment (rrichardson) [inline src/appendbuf.rs]: Good point. I'll add a guard

File: src/impls.rs (hunk 9 above)
Comment (rrichardson) [inline src/impls.rs]: the [] to express as_slice() is still unstable.

File: src/iobuf.rs (hunk 12 above)
Comment (rrichardson) [inline src/iobuf.rs]: Normally, I don't think so, but I think there was a parsing error without them in this case, can't say why.

File: src/iobuf.rs (hunk 13 above)
Comment (rrichardson) [inline src/iobuf.rs]: It's the British speelling.

File: src/iobuf.rs (hunk 14 above)
Comment (rrichardson) [inline src/iobuf.rs]: Ha.   sum is part of a deprecated library. I'm not sure if they plan to replace the lib or move sum.  Either way I'm pretty sure fold isn't going anywhere for a while.

File: src/appendbuf.rs (hunk 7 above)
Comment (rrichardson) [inline src/appendbuf.rs]: I should have paid more attention. It turns out there is no inclusive range operator :/

**Comment by cgaebel** [comment]
It lives as iobuf v5.2.1!

**Pull request merged**